src/main/java/io/github/hapjava/server/impl/pairing/PairingUpdateController.java

```diff
@@ -22,19 +24,45 @@ public HttpResponse handle(HttpRequest request) throws IOException {
     DecodeResult d = TypeLengthValueUtils.decode(request.getBody());
 
     int method = d.getByte(MessageType.METHOD);
-    if (method == 3) { // Add pairing
+    if (method == PairingMethod.ADD_PAIRING.getValue()) {
       byte[] username = d.getBytes(MessageType.USERNAME);
       byte[] ltpk = d.getBytes(MessageType.PUBLIC_KEY);
-      authInfo.createUser(authInfo.getMac() + new String(username, StandardCharsets.UTF_8), ltpk);
-    } else if (method == 4) { // Remove pairing
+      byte permissions = d.getByte(MessageType.PERMISSIONS);
+      authInfo.createUser(
+          authInfo.getMac() + new String(username, StandardCharsets.UTF_8), ltpk, permissions == 1);
+    } else if (method == PairingMethod.REMOVE_PAIRING.getValue()) {
       byte[] username = d.getBytes(MessageType.USERNAME);
       authInfo.removeUser(authInfo.getMac() + new String(username, StandardCharsets.UTF_8));
       if (!authInfo.hasUser()) {
         advertiser.setDiscoverable(true);
       }
+    } else if (method == PairingMethod.LIST_PAIRINGS.getValue()) {
+      TypeLengthValueUtils.Encoder e = TypeLengthValueUtils.getEncoder();
+
+      Collection<String> usernames = authInfo.listUsers();
+      boolean first = true;
+      Iterator<String> iterator = usernames.iterator();
+      while (iterator.hasNext()) {
+        String username = iterator.next();
+        if (first) {
+          e.add(MessageType.STATE, (byte) 2);
+          first = false;
+        } else {
+          e.add(MessageType.SEPARATOR);
+        }
+        e.add(MessageType.USERNAME, username);
```

Comment:
There's a bug here I realized as I was thinking about this last night while doing other things. We need to strip the MAC off the username we stored before returning it to iOS.

Reply:
Right, missed that one as well.

Reply:
probably at some point we should think about redesign of AuthInfo to store mac and user id separate. but it would be a breaking change..